### PR TITLE
Tracing: Use common traceID context value for opentracing and opentelemetry

### PR DESCRIPTION
**Pull request by zserge**

### CODE REVIEW

**Comment by zhichli** [comment]
@zserge could you please let me know once you have updated this branch with your latest change ?

**Comment by zhichli** [comment]
@zserge I have two asks based on my own verification of your latest changes. 
1. I found that for `400BadRequest` response got from check health api calls, there is no `traceID` attached to the response JSON body, which is actually due to the fact that `response.JSON()` is used to craft the `400BadGateway` response instead of `response.Error()` (see [line 558 in datasources.go](https://github.com/grafana/grafana/blob/dde0b93cf160fc97bce00645d4fe0063c34a2517/pkg/api/datasources.go#L558)). Is this implemented by design or should it be fixed by switching to `response.Error()`, which would trigger the injection of `traceID` in your current PR changes ? IMO, we might want to use `.Error()` call to be able to inject the `traceID` for easy tracing diagnostics since this scenario could potentially result in a case impacting user that requires our livesite investigation.
For example, when user accidentally put a wrong server endpoint (but they believe it's correct!) in a data source setting, and hit `Test` button which basically calls check health api, the response would have no `traceID` since the response status code is `400 bad gateway`. And user got confused and blocked, so we need to investigate such scenarios, which an injected `traceID` could potentially help.
image

3. Another case is that we want to have `traceID` in all datasource proxy's response body (at least the ones with errors) that can be captured in browser's network stack. But I found that even with your PR changes, the datasource proxy responses are not hitting the changes in your PR and thus have no `traceID` injected in the response body. Based on my investigation on the source code, it's due to the fact that the [datasource proxy request handling path](https://github.com/grafana/grafana/blob/dde0b93cf160fc97bce00645d4fe0063c34a2517/pkg/api/pluginproxy/ds_proxy.go#L112) is different from the [routing.Wrap() handler path](https://github.com/grafana/grafana/blob/dde0b93cf160fc97bce00645d4fe0063c34a2517/pkg/api/routing/routing.go#L15), which your PR changes reside. The datasource proxy handler does not [modify the response](https://github.com/grafana/grafana/blob/dde0b93cf160fc97bce00645d4fe0063c34a2517/pkg/api/pluginproxy/ds_proxy.go#L134) by injecting traceID in it, and even there is [no traceID extracted from the context and attached to the `data-proxy-log` logger](https://github.com/grafana/grafana/blob/dde0b93cf160fc97bce00645d4fe0063c34a2517/pkg/api/pluginproxy/ds_proxy.go#L118) (which is another issue I actually opened separately: #47365 ). So could we consider injecting traceID into all datasource proxy's response body which would greatly help us investigate many issues involving upstream plugins ? 

cc @taleena

**Comment by brahmnes** [comment]
@zserge , wouldn't it be ideal to have the trace id added as an http response header regardless of success status? Ie. it would be there whether it failed or not? Having it in the header also makes it easier to parse and not interfere with the response content if the request succeeded.

File: pkg/infra/tracing/tracing.go [hunk 1]
```diff
@@ -52,6 +53,21 @@ func ProvideService(cfg *setting.Cfg) (Tracer, error) {
 	return ots, ots.initOpentelemetryTracer()
 }
 
+type traceKey struct{}
+type traceValue struct {
+	ID        string
+	IsSampled bool
+}
+
+func TraceIDFromContext(c context.Context, sampled bool) string {
```
Comment (sakjur) [inline pkg/infra/tracing/tracing.go:62]:
```suggestion
func TraceIDFromContext(c context.Context, requireSampled bool) string {
```

I think `sampled` is a bit ambiguous here, not sure what to call it instead to make it clear that this is a filtering argument and not whether or not the trace is sampled. WDYT?

**Review by sakjur — APPROVED** [review]

**Review by idafurjes — APPROVED** [review]

**Pull request merged**